Refactor serialize

kaw2k opened the pull request:
## Things changed
- Our serialize is simpler(?)
- We renamed the prop `circular` on form to `addValidationFieldErrors`. Now by default validators only get `value, fieldValues`. If you want access to `fieldErrors` in your validators you need to set that flag to true on your form. When it is true, we serialize the form until it is consistent. 
## General approach:

Here we introduce a `tree` type which has a two properties (`ref` which is really value, and `refs` which is really children). With these types we implement a few methods like map and reduce. Since these methods know about the type they are in, they handle the recursion for us. This means we can map over our data structure just like we would map over an array and not worry about the details. 

More interesting is the use of `extend` and `extract`. Together, these functions form a comonad, a scary way to say that some values of the tree are dependent on the values around it. In this case, when generating errors, inputs can depend on other inputs near them in the tree. [Here](http://joneshf.github.io/programming/2015/12/31/Comonads-Monoids-and-Trees.html#comonad) is what inspired me to go down this route. 
## Room for improvement:
- The logic to convert children to tree nodes (found [here](https://github.com/willowtreeapps/react-formable/compare/refactor-serialize?expand=1#diff-62acef2590804164cb77b4b5aced00ebR66) and [here](https://github.com/willowtreeapps/react-formable/compare/refactor-serialize?expand=1#diff-8389c7b33442fb3f5cab4ca39fd8be63R23) could / should be abstracted away.
- We are using plain javascript objects to represent our type [here](https://github.com/willowtreeapps/react-formable/compare/refactor-serialize?expand=1#diff-c749ff30aede17a22cf83e1cb3bf7dadR21). This means that the methods attached to the tree type are created for each node. If we really care we can make a proper class and attach them to the prototype
- The values `ref` and `refs` from `getInputs` as well as the constructor of the tree could be improved / renamed. `value` and `children` would be more explicit.
- Break out the functions in the tree file
- Tests...

kaw2k committed:
Adding a tree type

The tree type contains two properties “ref” and “refs” (or val and
children). We probably should rename these.

map, extract, extend, and reduce are methods that give our tree
additional powers.

kaw2k committed:
Updating formable to use new tree type

kaw2k committed:
Renaming props.circular to props.addValidationErrors

oconn commented on src/helpers/tree.js:
```diff
@@ -0,0 +1,48 @@
+    return obj.extract();
+}
+
+export default function tree({ ref, refs }) {
```
@kaw2k I really like how compact these methods are. My one comment would be, for readability / digestibility purposes, it would probably be nice to have signatures above these methods.

kaw2k replied:
Thanks @oconn, done!

kaw2k committed:
Documenting tree

kaw2k committed:
Updating docs to reflect new changes

kaw2k committed:
Fixing an issue where null values were coming through validators

kaw2k commented:
Ok, this should be good to go.

uttrasey commented:
Looks good to me :+1:

kaw2k committed:
Moving from forked gulp task to real gulp task

kaw2k committed:
Fixing package.json errors

kaw2k committed:
Fixing importing files :(

kaw2k committed:
Changing how trees are constructed

Before, trees accept an object with ref and refs. This commit removes
the object and renames refs -> value and refs -> children.

The impetus to rename the parameters stemmed from making them actually
match what the tree expects. Having them as ref and refs tightly
coupled the tree to react.

The impetus for removing the object is just to clean up noise.

I am open to change in any way

kaw2k committed:
Fixing fixing build / tests
